Test/WhatsNewDataModelTests.swift

```
@@ -15,10 +15,10 @@ class WhatsNewDataModelTests: XCTestCase {
     func testParsing() {
```

While the tests in this file seem good, they are missing some basic format tests e.g. We should verify 
1 - that the root element of the JSON file is an array.
2 - that the value of `version` key in each item of the root array is correctly parsed into our `Version` model.
3 - that values of the `platforms` arrays aren't anything other than `android` and/or `ios`.
4 - that values of the `image` attribute is of the format `screen_{only_a_positive_number}.png`.

---

> 1 - that the root element of the JSON file is an array.
If root element is not an array, then there will be no parsed data to test because the code is expecting an array and used that array to parse data. 
> 2 - that the value of version key in each item of the root array is correctly parsed into our Version model.
Good point.  
> 3 - that values of the platforms arrays aren't anything other than Android and/or ios.
The array of `platforms`  was added so that in case of new client it can be used. If we bound it to iOS and Android then tests needs to be updated whenever a new client is added. 
> 4 - that values of the image attribute is of the format screen_{only_a_positive_number}.png.
`image` can be named in any way. There isn't any hard and fast rule for `image` naming.

---

> If root element is not an array, then there will be no parsed data to test because the code is expecting an array and used that array to parse data.

Good enuff.

> The array of platforms was added so that in case of new client it can be used. If we bound it to iOS and Android then tests needs to be updated whenever a new client is added.

Exactly! Tests should fail whenever a new/unknown platform is added. Also, it would ensure that there's no type in the platform values (since they are predefined values).

> image can be named in any way. There isn't any hard and fast rule for image naming.

On Android, it is. Would be good if iOS does this too.

---

> Exactly! Tests should fail whenever a new/unknown platform is added. Also, it would ensure that there's no typo in the platform values (since they are predefined values).

Oh, unfortunately, iOS only consider messages those are intended for iOS. The code will ignore any message if platforms don't have element `iOS`. 

> On Android, it is. Would be good if iOS does this too.

But I guess anyone can name `image` like `whats_new_image_due_date` or `whats_new_image_navigation_changes`. I hope by naming images like this wouldn't break whatsNew flow. Right?

---

> Oh, unfortunately, iOS only consider messages those are intended for iOS. The code will ignore any message if platforms don't have element iOS.

What if there's a typo in the file i.e. instead of `iOS` there `IoS` or `IOS` or `i0s` (i zero s) ?

> But I guess anyone can name image like whats_new_image_due_date or whats_new_image_navigation_changes. I hope by naming images like this wouldn't break whatsNew flow. Right?

As long as the value of `image` attribute translates to an actual file, it won't be a problem. But as I said it would be a *good to have* thing on iOS too.

---

> What if there's a typo in the file i.e. instead of iOS there is IoS or IOS or i0s (i zero s) ?

`iOS`, `IoS` and `IOS` will be lowered case and then compared to hardcoded `ios`. Zero is not a valid value so it will be ignored.

---

And that's exactly why I mentioned that test!

---

> And that's exactly why I mentioned that test!

This can only be tested with test data, not with actual data.

---

For the sake of checking if the tests are written correctly, you can modify the original JSON file and run your tests.
In future, whenever the JSON will be updated for a specific version, the tests will run and if there's any typo or mistake the tests will fail. Which is inherently the basic purpose of having these/any tests.

---

> 2 - that the value of version key in each item of the root array is correctly parsed into our Version model.

If version string will not be converted to [Version: https://github.com/edx/edx-app-ios/blob/saeed/LEARNER-2527/Source/WhatsNewDataModel.swift#L87] then there will be no data returned by the WhatsNewDataModel.

---

@miankhalid said changeset is beyond its scope. This can be discussed and new tickets can be created under the scope of https://openedx.atlassian.net/browse/LEARNER-2625

---

Deferring this to   [LEARNER-2625](https://openedx.atlassian.net/browse/LEARNER-2625).